eval/src/vespa/eval/instruction/join_with_number_function.cpp

```diff
@@ -14,39 +14,55 @@ namespace vespalib::eval {
 
 using Instruction = InterpretedFunction::Instruction;
 using State = InterpretedFunction::State;
+using vespalib::eval::tensor_function::unwrap_param;
+using vespalib::eval::tensor_function::wrap_param;
 
 namespace {
 
-template <typename CT, bool inplace>
-ArrayRef<CT> make_dst_cells(ConstArrayRef<CT> src_cells, Stash &stash) {
-    if (inplace) {
+struct JoinWithNumberParam {
+    const ValueType res_type;
+    const join_fun_t function;
+    JoinWithNumberParam(const ValueType &r, join_fun_t f) : res_type(r), function(f) {}
+};
+
+template <typename ICT, typename OCT, bool inplace>
+ArrayRef<OCT> make_dst_cells(ConstArrayRef<ICT> src_cells, Stash &stash) {
+    if constexpr (inplace) {
+        static_assert(std::is_same_v<ICT,OCT>);
         return unconstify(src_cells);
     } else {
-        return stash.create_uninitialized_array<CT>(src_cells.size());
+        return stash.create_uninitialized_array<OCT>(src_cells.size());
     }
 }
 
-template <typename CT, typename Fun, bool inplace, bool swap>
-void my_number_join_op(State &state, uint64_t param) {
+template <typename ICT, typename OCT, typename Fun, bool inplace, bool swap>
+void my_number_join_op(State &state, uint64_t param_in) {
+    const auto &param = unwrap_param<JoinWithNumberParam>(param_in);
     using OP = typename std::conditional<swap,SwapArgs2<Fun>,Fun>::type;
-    OP my_op((join_fun_t)param);
+    OP my_op(param.function);
     const Value &tensor = state.peek(swap ? 0 : 1);
-    CT number = state.peek(swap ? 1 : 0).as_double();
-    auto src_cells = tensor.cells().typify<CT>();
-    auto dst_cells = make_dst_cells<CT, inplace>(src_cells, state.stash);
+    OCT number = state.peek(swap ? 1 : 0).as_double();
+    auto src_cells = tensor.cells().typify<ICT>();
+    auto dst_cells = make_dst_cells<ICT, OCT, inplace>(src_cells, state.stash);
     apply_op2_vec_num(dst_cells.begin(), src_cells.begin(), number, dst_cells.size(), my_op);
     if (inplace) {
         state.pop_pop_push(tensor);
     } else {
-        state.pop_pop_push(state.stash.create<ValueView>(tensor.type(), tensor.index(), TypedCells(dst_cells)));
+        state.pop_pop_push(state.stash.create<ValueView>(param.res_type, tensor.index(), TypedCells(dst_cells)));
     }
 }
 
 struct SelectJoinWithNumberOp {
-    template<typename CT, typename Fun,
-             typename InputIsMutable, typename NumberWasLeft>
+    template<typename CM, typename Fun,
+             typename Inplace, typename NumberWasLeft>
     static auto invoke() {
-        return my_number_join_op<CT, Fun, InputIsMutable::value, NumberWasLeft::value>;
+        constexpr CellMeta icm = CM::value;
+        constexpr CellMeta num(CellType::DOUBLE, true);
+        constexpr CellMeta ocm = CellMeta::join(icm, num); 
+        using ICT = CellValueType<icm.cell_type>;
+        using OCT = CellValueType<ocm.cell_type>;
+        constexpr bool inplace = (Inplace::value && std::is_same_v<ICT,OCT>);
```

Comment:
this makes it look like Inplace::value is not authorative. I suggest adding a non-static assert checking that Inplace::value implies ICT and OCT being the same type; to trap the generated code that will never be called instead of handling parameter inconsistencies silently.

Reply:
an alternative is to downgrade the flag from 'inplace' to 'pri_mut' and infer inplace as done here. That will reduce unused code generation.